selection.go

```diff
@@ -0,0 +1,267 @@
+package ice
+
+import (
+	"bytes"
+	"sort"
+	"net"
+	"time"
+
+	"github.com/pion/stun"
+)
+
+type nominatedPair struct {
```

Comment:
I would drop the dedicated `nominatedPair` structure, and maybe mark our `validPairs` as nominated (maybe?)

Thinking of your comments from earlier
```
a) check the transaction without worring about the pendingTransaction structure (because we shouldn't risk to loose this tx in case of overflow)
b) In case we make a nomination in the controlling side we need to differentiate between a normal success response and a response to a nomination (in the later we select the pair)
b) In case we get a nomination in the controlled side for a pair that is not validate yet, we need to start another cycle of request/response and then answer the nomination
```

a) I think it is ok to loose the tx, we will keep sending the binding-success
b) I would add another field to transaction something like `useCandidateResponse`, keep the code simpler
c) If we get a `useCandidate` and the pair isn't valid I think we will need to use if (if we are controlled) if we are dealing with aggressive remotes.

Reply:
Can you expand on your first point? This is something I'm not sure how to do, we need to *always* reply to a nomination, but I'm not sure how to do that. It seems to me we cannot loose that tx and we will need to reply even if there if is not a valid pair and we need to perform a new request/response cycle

Reply:
https://tools.ietf.org/html/rfc8445#section-7.3.1.5

If the controlled agent does not accept the request from the
   controlling agent, the controlled agent MUST reject the nomination
   request with an appropriate error code response (e.g., 400)
   [RFC5389].

Reply:
We need to call `nominatePair` in a loop, since the connection may be lossy our original might get lost.

I would entire failure by just having a timer that records how long we have been in checking (if we are stuck in checking for more then 30 seconds we should go to fail)

I would update `taskLoop` to call the `ControllingSelector`, with logic like this
* If we have a `selectedPair` ping it
* If we have a CandidatePair that is nominated sending a Request with `useCandidate`
* If we don't have any nominated CandidatePairs ping everyone

----

To distinguish if the response is for a `useCandidate` or an old transaction I would update like this
```
type bindingRequest struct {
    transactionID []byte
    destination   net.Addr
    isUseCandidate bool
}
```

and if we get a response back for `isUseCandidate == true` this becomes our selected pair (on the controlling side)

Reply:
I did this change, however, I feel a bit unsure about this part of the standard

```
-       // If the controlled agent does not accept the request from the
-       // controlling agent, the controlled agent MUST reject the nomination
-       // request with an appropriate error code response (e.g., 400)
-       // [RFC5389].
```

in pion<->pion this would work because of the general timeout, but not sure when interacting with other implementations